crates/bevy_pbr/src/material_bind_groups.rs

```diff
@@ -155,634 +276,1150 @@ impl From<u32> for MaterialBindGroupIndex {
-                material_bindless_bind_group.allocate()
-            }
-            MaterialBindGroup::NonBindless(ref mut material_non_bindless_bind_group) => {
-                material_non_bindless_bind_group.allocate()
-            }
+        MaterialBindlessIndexTable {
+            buffer,
+            buffer_dirty: BufferDirtyState::NeedsReserve,
+            phantom: PhantomData,
         }
     }
 
-    /// Assigns an unprepared bind group to the group and slot specified in the
-    /// [`MaterialBindingId`].
-    fn init(
+    /// Returns the binding index table for a single material.
+    ///
+    /// Element *i* of the returned binding index table contains the slot of the
+    /// bindless resource with bindless index *i*.
+    fn get(&self, slot: MaterialBindGroupSlot, bindless_descriptor: &BindlessDescriptor) -> &[u32] {
+        let struct_size = bindless_descriptor.resources.len();
+        let start = struct_size * slot.0 as usize;
+        &self.buffer.values()[start..(start + struct_size)]
+    }
+
+    /// Updates the binding index table for a single material.
+    ///
+    /// The `allocated_resource_slots` map contains a mapping from the
+    /// [`BindlessIndex`] of each resource that the material references to the
+    /// slot that that resource occupies in the appropriate binding array. This
+    /// method serializes that map into a binding index table that the shader
+    /// can read.
+    fn set(
         &mut self,
-        render_device: &RenderDevice,
-        bind_group_layout: &BindGroupLayout,
         slot: MaterialBindGroupSlot,
-        unprepared_bind_group: UnpreparedBindGroup<M::Data>,
+        allocated_resource_slots: &HashMap<BindlessIndex, u32>,
+        bindless_descriptor: &BindlessDescriptor,
     ) {
-        match *self {
-            MaterialBindGroup::Bindless(ref mut material_bindless_bind_group) => {
-                material_bindless_bind_group.init(
-                    render_device,
-                    bind_group_layout,
-                    slot,
-                    unprepared_bind_group,
-                );
-            }
-            MaterialBindGroup::NonBindless(ref mut material_non_bindless_bind_group) => {
-                material_non_bindless_bind_group.init(
-                    render_device,
-                    bind_group_layout,
-                    slot,
-                    unprepared_bind_group,
-                );
-            }
+        let table_len = bindless_descriptor.resources.len();
+        let range = (slot.0 as usize * table_len)..((slot.0 as usize + 1) * table_len);
+        while self.buffer.len() < range.end {
+            self.buffer.push(0);
+        }
+
+        for (&bindless_index, &resource_slot) in allocated_resource_slots {
+            self.buffer
+                .set(*bindless_index + range.start as u32, resource_slot);
         }
+
+        // Mark the buffer as needing to be recreated, in case we grew it.
```

Comment:
is it possible to detect whether the buffer grew (by saving capacity before and comparing after) and only mark it if so?

Reply:
Yes, but I'd like to do this in a follow up because the optimal is actually more complex than that: we can actually avoid recreating the bind group at all.